src/utils/token_to_fiat_oracle.ts

```diff
@@ -0,0 +1,137 @@
+ * 2. constructor(private readonly client = new CoinGecko()) {}
+ * 3. const response = await this.client.ping() // for the ping() function
+ *    const response = await this.client.simple.price(params) // for getting the price of a currency
+ */
+
+export class TokenToFiatOracle implements FiatOracle {
+	private cachedPrices: TokenFiatRate[] = [];
+	private cacheTimestamp = 0;
+	private syncPromise?: Promise<void>;
+
+	/**
+	 * @param {FiatID[]} fiats The currency IDs to quote tokens in
+	 * @param {TokenID[]} tokens The token ID of the requested coin
+	 * @param {number} cacheLifespan Milliseconds that has to pass until the cache is invalid
+	 * @param {@link fetch} _fetch The standard fetch method. Defined here for testing propuses
+	 * @return {TokenToFiatOracle} A class representing the oracle for the selected currencies
+	 */
+	constructor(
+		private readonly fiats: FiatID[] = ['USD'],
+		private readonly tokens: TokenID[] = ['ARWEAVE'],
+		private readonly cacheLifespan = pollingIntervalMilliseconds,
+		private readonly _fetch = fetch
+	) {}
+
+	private get currentlyFetchingPrice(): boolean {
+		return !!this.syncPromise;
+	}
+
+	private get tokensCSV(): string {
+		return this.tokens.map(tokenIdToThirdParty).join(',');
+	}
+
+	private get fiatCSV(): string {
+		return this.fiats.map(fiatIdToThirdParty).join(',');
+	}
+
+	private get shouldRefreshCacheData(): boolean {
+		const currentTimestamp = Date.now();
+		const deltaTimestamp = currentTimestamp - this.cacheTimestamp;
+		return deltaTimestamp >= this.cacheLifespan;
+	}
+
+	/**
+	 * @param {FiatID} fiat The fiat currency ID
+	 * @param {TokenID} token The token ID
+	 * @throws {@link Error} If no such pair cached
+	 * @returns {Promise<TokenToFiatPrice>} The cached price value
+	 */
+	public async getPriceForFiatTokenPair(pair: TokenFiatPair): Promise<TokenFiatRate> {
+		await this.checkCache();
+		const pricePair = this.cachedPrices.find(cachePairFilterFactory(pair));
+		if (!pricePair) {
+			throw new Error(`No such pair (${pair.fiat}, ${pair.token})`);
+		}
+		return pricePair;
+	}
+
+	private async checkCache(): Promise<void> {
+		if (this.currentlyFetchingPrice) {
+			return this.syncPromise;
+		}
+		if (this.shouldRefreshCacheData) {
+			this.syncPromise = this.fetchPrices();
+			await this.syncPromise;
+			delete this.syncPromise;
+		}
+	}
+
+	private async fetchPrices(): Promise<void> {
+		const queryUrl = this.getQueryRequestUrl();
+		const fetchResponse = await this._fetch(queryUrl);
+		const responseData: CoinGeckoPriceResponseData = await fetchResponse.json();
+		this.fiats.forEach((fiat) => {
+			this.tokens.forEach((token) => {
+				const priceValue = responseData[token][fiat];
+				this.updateCachePair(new TokenFiatPair(fiat, token), priceValue);
+			});
+		});
+	}
+
+	private updateCachePair(pair: TokenFiatPair, rate: number): void {
+		/* It is safe to pop, then push the element as javascript is single-threaded */
+		try {
+			this.popSingleCachePair(pair);
```

Comment:
Keep running into errors here on the first run through. I don't think the case when the cachedPrices array is initially empty has been accounted for here. We could bring the function below's logic inside here and clean this section up:

```ts
private updateCachePair(pair: TokenFiatPair, rate: number): void {
		// Check for cached token fiat pair
		const cachePairIndex = this.cachedPrices.findIndex((p) => p.fiat === pair.fiat && p.token === pair.token);

		if (cachePairIndex === -1) {
			// If found, remove from cache
			/* It is safe to pop, then push the element as javascript is single-threaded */
			this.cachedPrices.splice(cachePairIndex, 1)[0]; // popSingleCachePair
		}

		// Add new token fiat pair
		const updatedPrice = new TokenFiatRate(pair.fiat, pair.token, rate);
		this.cachedPrices.push(updatedPrice);
	}
```

Reply:
The oracle is no longer dealing with multiple currencies, but instead, it's now referred to as a specific and immutable pair